lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2522,66 +2538,61 @@ namespace {
-            addDeclsReversed(classTemplate->specializations());
-            continue;
-          }
+      Impl.ImportedDecls[{decl, getVersion()}] = extension;
+
+      // Insert these backwards into "namespaceDecls" so we can pop them off
+      // the end without loosing order.
+      SmallVector<clang::Decl *, 16> namespaceDecls;
+      auto addDeclsReversed = [&](auto decls) {
+        auto begin = decls.begin();
+        auto end = decls.end();
+        int currentSize = namespaceDecls.size();
+        int declCount = currentSize + std::distance(begin, end);
+        namespaceDecls.resize(declCount);
+        for (int index = declCount - 1; index >= currentSize; --index)
+          namespaceDecls[index] = *(begin++);
+      };
+      addDeclsReversed(decl->decls());
+      while (!namespaceDecls.empty()) {
+        auto nd = dyn_cast<clang::NamedDecl>(namespaceDecls.pop_back_val());
+        // Make sure we only import the defenition of a record.
+        if (auto tagDecl = dyn_cast_or_null<clang::TagDecl>(nd))
+          // Some decls, for example ClassTemplateSpecializationDecls, won't
+          // have a definition here. That's OK.
+          nd = tagDecl->getDefinition() ? tagDecl->getDefinition() : tagDecl;
+        if (!nd)
+          continue;
 
-          auto member = Impl.importDecl(nd, getVersion());
-          if (!member || addedMembers.count(member) ||
-              isa<clang::NamespaceDecl>(nd))
-            continue;
-          // This happens (for example) when a struct is declared inside another
-          // struct inside a namespace but defined out of line.
-          assert(member->getDeclContext()->getAsDecl());
-          if (dyn_cast<ExtensionDecl>(member->getDeclContext()->getAsDecl()) !=
-              extension)
-            continue;
-          extension->addMember(member);
-          addedMembers.insert(member);
+        // Special case class templates: import all their specilizations here.
+        if (auto classTemplate = dyn_cast<clang::ClassTemplateDecl>(nd)) {
+          addDeclsReversed(classTemplate->specializations());
+          continue;
         }
+
+        bool useCanonicalDecl = !isa<clang::NamespaceDecl>(nd);
+        auto member = Impl.importDecl(nd, getVersion(), useCanonicalDecl);
+        if (!member || addedMembers.count(member) ||
+            isa<clang::NamespaceDecl>(nd))
+          continue;
+        // This happens (for example) when a struct is declared inside another
+        // struct inside a namespace but defined out of line.
+        assert(member->getDeclContext()->getAsDecl());
+        if (dyn_cast<ExtensionDecl>(member->getDeclContext()->getAsDecl()) !=
+            extension)
+          continue;
+        extension->addMember(member);
+        addedMembers.insert(member);
       }
 
-      if (!extension->getMembers().empty())
-        enumDecl->addExtension(extension);
+      enumDecl->addExtension(extension);
```

Comment:
We don't really want empty extensions. Is there a reason you removed this?

Reply:
Previously we were returning an enum from `VisitNamespaceDecl` and we could skip empty extensions, but now we're returning an extension which should be in a valid state (i.e. have a known extended type) even if it's empty.
We could return `nullptr` for empty namespace decls, but I think that would mess with import of nested namespaces (when a namespace has no other decls in it except another namespace).
Do we need to try harder & keep purging the empty extensions somehow?

Reply:
Thanks for explaining. In that case, this is fine.